torchvision/prototype/datasets/_builtin/lsun.py

```diff
@@ -0,0 +1,186 @@
+import functools
+import io
+import pathlib
+import re
+from typing import Any, Callable, Dict, List, Optional, Tuple, Iterator
+
+import torch
+from torchdata.datapipes.iter import IterDataPipe, Mapper, OnDiskCacheHolder, Concater, IterableWrapper
+from torchvision.prototype.datasets.utils import (
+    Dataset,
+    DatasetConfig,
+    DatasetInfo,
+    HttpResource,
+    OnlineResource,
+    DatasetType,
+)
+from torchvision.prototype.datasets.utils._internal import hint_sharding, hint_shuffling
+from torchvision.prototype.features import Label
+
+# We need lmdb.Environment as annotation, but lmdb is an optional requirement at import
+try:
+    import lmdb
+
+    Environment = lmdb.Environment
+except ImportError:
+    Environment = Any
+
+
+class LmdbKeyExtractor(IterDataPipe[Tuple[str, bytes]]):
+    def __init__(self, datapipe: IterDataPipe[str]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[str, bytes]]:
+        import lmdb
+
+        for path in self.datapipe:
+            with lmdb.open(path, readonly=True) as env:
+                with env.begin(write=False) as txn:
+                    keys = b"\n".join(key for key in txn.cursor().iternext(keys=True, values=False))
+                    yield path, keys
+
+
+class LmdbLoader(IterDataPipe[Tuple[Environment, bytes]]):
+    def __init__(self, datapipe: IterDataPipe[str]) -> None:
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[Environment, bytes]]:  # type: ignore[valid-type]
+        import lmdb
+
+        for cache_path in self.datapipe:
+            env = lmdb.open(str(pathlib.Path(cache_path).parent), readonly=True)
+
+            with open(cache_path, "rb") as file:
+                for key in file:
+                    yield env, key.strip()
+
+
+class LmdbReader(IterDataPipe):
+    def __init__(self, datapipe: IterDataPipe[Tuple[Environment, bytes]]):
+        self.datapipe = datapipe
+
+    def __iter__(self) -> Iterator[Tuple[str, bytes, io.BytesIO]]:
+        for env, key in self.datapipe:
+            with env.begin(write=False) as txn:
+                yield env.path(), key, io.BytesIO(txn.get(key))
+
+
+class LsunHttpResource(HttpResource):
+    def __init__(self, *args: Any, extract: bool = True, **kwargs: Any) -> None:
+        super().__init__(*args, extract=extract, **kwargs)
+
+    def _loader(self, path: pathlib.Path) -> IterDataPipe[str]:
+        # LMDB datasets cannot be loaded through an open file handle, but have to be loaded through the path of the
+        # parent directory.
+        return IterableWrapper([str(next(path.rglob("data.mdb")).parent)])
```

Comment:
@NicolasHug this would have been supported with the `loader` parameter we removed in #5282. But I guess since this is (currently) the only case that needs this, we can special case it.